{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "62f67447-1520-4aa1-934d-c0a68571f54c",
   "metadata": {},
   "source": [
    "# Additional datasets from other studies\n",
    "This notebook describes the methods to access the genomes from other studies used in the analysis\n",
    "\n",
    "## Table of Contents\n",
    "* [Dataset from Bickhart et al (695 MAGs, 286 HQ, 284 Bacteria)](#bickhart)\n",
    "* [Dataset from Liu et al (557 MAGs, 153 HQ, 153 Bacteria)](#liu)\n",
    "* [Dataset from Christoph et al (376 MAGs, 73 HQ, 73 Bacteria)](#christoph)\n",
    "* [Dataset from Sharrar et al (1334 MAGs, 374 HQ, 350 Bacteria)](#sharrar)\n",
    "    * Sanitize Contig Names\n",
    "* [Dataset from Chen et al (96 MAGs, 24 HQ, 21 Bacteria)](#chen)\n",
    "    * Sanitize Contig Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9225c0c3-6d49-4a8c-85ee-d44a0966ef2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please install this libraries first\n",
    "import pandas as pd\n",
    "import json\n",
    "from pathlib import Path\n",
    "import tarfile\n",
    "import gzip\n",
    "import shutil\n",
    "import requests, json\n",
    "import ast\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "209fa532-129a-4f7d-a027-aba27fa6b620",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up paths\n",
    "## Path to BGCflow directory\n",
    "external_data_path = Path(\"../../bgcflow/data/external\")\n",
    "\n",
    "## Path to tables from other studies\n",
    "external_table_path = Path(\"../tables/other_studies_HQ\")\n",
    "external_table_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "## Generate a config template for BGCflow\n",
    "df_bgcflow_template = pd.DataFrame(columns=[\"source\",\"organism\",\"genus\",\"species\",\"strain\",\"closest_placement_reference\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "42a3ec64-5788-44da-b4bb-d0edd9d51523",
   "metadata": {},
   "outputs": [],
   "source": [
    "# external files required to run the notebook\n",
    "# need to make this available in github!\n",
    "\n",
    "# Not obvious how to fetch with wget\n",
    "file_christoph_ncbi = Path(\"../data/supplementary_tables/PRJNA449266_AssemblyDetails.txt\")\n",
    "file_chen_metadata = Path(\"../data/supplementary_tables/Table_1_Discovery of an Abundance of Biosynthetic Gene Clusters in Shark Bay Microbial Mats.XLSX\")\n",
    "\n",
    "# Manual Run - GTDB runs v202\n",
    "file_gtdb = Path(\"../data/pre_analysis_tables/gtdb.all_hqmags_studies.tsv\")\n",
    "file_gtdb_warning = Path(\"../data/pre_analysis_tables/gtdbtk.warnings.log\")\n",
    "\n",
    "# Manual Run - CheckM\n",
    "file_bickhart_checkm = Path(\"../data/pre_analysis_tables/Bickhart_et_al_checkm_out.tsv\") # CheckM runs of all bins\n",
    "file_chen_checkm = Path(\"../data/pre_analysis_tables/Chen_et_al_df_checkm_stats.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "58902dd5-0107-4836-96f1-868b45babcb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GTDB\n",
    "df_gtdb = pd.read_csv(file_gtdb, sep=\"\\t\")\n",
    "df_gtdb = df_gtdb.set_index(\"user_genome\", drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "edb2693c-65a6-40ad-9c12-99f277ff79a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>user_genome</th>\n",
       "      <th>original_user_genome</th>\n",
       "      <th>classification</th>\n",
       "      <th>fastani_reference</th>\n",
       "      <th>fastani_reference_radius</th>\n",
       "      <th>fastani_taxonomy</th>\n",
       "      <th>fastani_ani</th>\n",
       "      <th>fastani_af</th>\n",
       "      <th>closest_placement_reference</th>\n",
       "      <th>closest_placement_radius</th>\n",
       "      <th>...</th>\n",
       "      <th>closest_placement_ani</th>\n",
       "      <th>closest_placement_af</th>\n",
       "      <th>pplacer_taxonomy</th>\n",
       "      <th>classification_method</th>\n",
       "      <th>note</th>\n",
       "      <th>other_related_references(genome_id,species_name,radius,ANI,AF)</th>\n",
       "      <th>msa_percent</th>\n",
       "      <th>translation_table</th>\n",
       "      <th>red_value</th>\n",
       "      <th>warnings</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>user_genome</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54_67_14</th>\n",
       "      <td>54_67_14</td>\n",
       "      <td>14_0903_02_20cm_Proteobacteria_54_67_14</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>GCA_005878835.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>99.99</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GCA_005878835.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>99.99</td>\n",
       "      <td>1.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>topological placement and ANI have congruent s...</td>\n",
       "      <td>GCA_005878095.1, s__13-2-20CM-66-19 sp00587809...</td>\n",
       "      <td>95.79</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159_36_9</th>\n",
       "      <td>159_36_9</td>\n",
       "      <td>14_0903_02_30cm_Rickettsiella_grylli_159_36_9</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>GCA_005877695.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>99.98</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GCA_005877695.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>99.98</td>\n",
       "      <td>1.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>topological placement and ANI have congruent s...</td>\n",
       "      <td>GCF_001881495.1, s__Rickettsiella_A isopodorum...</td>\n",
       "      <td>95.59</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157_38_13</th>\n",
       "      <td>157_38_13</td>\n",
       "      <td>14_0903_02_30cm_Sphingobacteriales_157_38_13</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>GCA_005882315.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GCA_005882315.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>topological placement and ANI have congruent s...</td>\n",
       "      <td>GCA_005882975.1, s__VBAS01 sp005882975, 95.0, ...</td>\n",
       "      <td>96.72</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165_43_8</th>\n",
       "      <td>165_43_8</td>\n",
       "      <td>14_0903_02_30cm_Sphingobacteriales_165_43_8</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>GCA_005884665.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GCA_005884665.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>topological placement and ANI have congruent s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88.70</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156_68_15</th>\n",
       "      <td>156_68_15</td>\n",
       "      <td>14_0903_02_30cm_Sphingomonadales_156_68_15</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Alphaproteoba...</td>\n",
       "      <td>GCA_005883305.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Alphaproteoba...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GCA_005883305.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Alphaproteoba...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>topological placement and ANI have congruent s...</td>\n",
       "      <td>GCF_007995055.1, s__Allosphingosinicella ginse...</td>\n",
       "      <td>97.00</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin3c.78</th>\n",
       "      <td>bin3c.78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002298075.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>...</td>\n",
       "      <td>78.08</td>\n",
       "      <td>0.1</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>taxonomic classification defined by topology a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.65</td>\n",
       "      <td>11</td>\n",
       "      <td>0.957236</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin3c.79</th>\n",
       "      <td>bin3c.79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>ANI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_900540045.1, s__Firm-11 sp900540045, 95.0,...</td>\n",
       "      <td>92.30</td>\n",
       "      <td>11</td>\n",
       "      <td>0.942301</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin3c.91</th>\n",
       "      <td>bin3c.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>taxonomic classification fully defined by topo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.33</td>\n",
       "      <td>11</td>\n",
       "      <td>0.864885</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin3c.92</th>\n",
       "      <td>bin3c.92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>taxonomic novelty determined using RED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94.92</td>\n",
       "      <td>11</td>\n",
       "      <td>0.894587</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin3c.98</th>\n",
       "      <td>bin3c.98</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...</td>\n",
       "      <td>taxonomic novelty determined using RED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.55</td>\n",
       "      <td>11</td>\n",
       "      <td>0.874257</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2453 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            user_genome                           original_user_genome  \\\n",
       "user_genome                                                              \n",
       "54_67_14       54_67_14        14_0903_02_20cm_Proteobacteria_54_67_14   \n",
       "159_36_9       159_36_9  14_0903_02_30cm_Rickettsiella_grylli_159_36_9   \n",
       "157_38_13     157_38_13   14_0903_02_30cm_Sphingobacteriales_157_38_13   \n",
       "165_43_8       165_43_8    14_0903_02_30cm_Sphingobacteriales_165_43_8   \n",
       "156_68_15     156_68_15     14_0903_02_30cm_Sphingomonadales_156_68_15   \n",
       "...                 ...                                            ...   \n",
       "bin3c.78       bin3c.78                                            NaN   \n",
       "bin3c.79       bin3c.79                                            NaN   \n",
       "bin3c.91       bin3c.91                                            NaN   \n",
       "bin3c.92       bin3c.92                                            NaN   \n",
       "bin3c.98       bin3c.98                                            NaN   \n",
       "\n",
       "                                                classification  \\\n",
       "user_genome                                                      \n",
       "54_67_14     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "159_36_9     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "157_38_13    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "165_43_8     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "156_68_15    d__Bacteria;p__Proteobacteria;c__Alphaproteoba...   \n",
       "...                                                        ...   \n",
       "bin3c.78     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "bin3c.79     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.91     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.92     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.98     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "\n",
       "            fastani_reference  fastani_reference_radius  \\\n",
       "user_genome                                               \n",
       "54_67_14      GCA_005878835.1                      95.0   \n",
       "159_36_9      GCA_005877695.1                      95.0   \n",
       "157_38_13     GCA_005882315.1                      95.0   \n",
       "165_43_8      GCA_005884665.1                      95.0   \n",
       "156_68_15     GCA_005883305.1                      95.0   \n",
       "...                       ...                       ...   \n",
       "bin3c.78                  NaN                       NaN   \n",
       "bin3c.79                  NaN                       NaN   \n",
       "bin3c.91                  NaN                       NaN   \n",
       "bin3c.92                  NaN                       NaN   \n",
       "bin3c.98                  NaN                       NaN   \n",
       "\n",
       "                                              fastani_taxonomy  fastani_ani  \\\n",
       "user_genome                                                                   \n",
       "54_67_14     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...        99.99   \n",
       "159_36_9     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...        99.98   \n",
       "157_38_13    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       100.00   \n",
       "165_43_8     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       100.00   \n",
       "156_68_15    d__Bacteria;p__Proteobacteria;c__Alphaproteoba...       100.00   \n",
       "...                                                        ...          ...   \n",
       "bin3c.78                                                   NaN          NaN   \n",
       "bin3c.79                                                   NaN          NaN   \n",
       "bin3c.91                                                   NaN          NaN   \n",
       "bin3c.92                                                   NaN          NaN   \n",
       "bin3c.98                                                   NaN          NaN   \n",
       "\n",
       "             fastani_af closest_placement_reference  closest_placement_radius  \\\n",
       "user_genome                                                                     \n",
       "54_67_14            1.0             GCA_005878835.1                      95.0   \n",
       "159_36_9            1.0             GCA_005877695.1                      95.0   \n",
       "157_38_13           1.0             GCA_005882315.1                      95.0   \n",
       "165_43_8            1.0             GCA_005884665.1                      95.0   \n",
       "156_68_15           1.0             GCA_005883305.1                      95.0   \n",
       "...                 ...                         ...                       ...   \n",
       "bin3c.78            NaN             GCA_002298075.1                      95.0   \n",
       "bin3c.79            NaN                         NaN                       NaN   \n",
       "bin3c.91            NaN                         NaN                       NaN   \n",
       "bin3c.92            NaN                         NaN                       NaN   \n",
       "bin3c.98            NaN                         NaN                       NaN   \n",
       "\n",
       "             ... closest_placement_ani  closest_placement_af  \\\n",
       "user_genome  ...                                               \n",
       "54_67_14     ...                 99.99                   1.0   \n",
       "159_36_9     ...                 99.98                   1.0   \n",
       "157_38_13    ...                100.00                   1.0   \n",
       "165_43_8     ...                100.00                   1.0   \n",
       "156_68_15    ...                100.00                   1.0   \n",
       "...          ...                   ...                   ...   \n",
       "bin3c.78     ...                 78.08                   0.1   \n",
       "bin3c.79     ...                   NaN                   NaN   \n",
       "bin3c.91     ...                   NaN                   NaN   \n",
       "bin3c.92     ...                   NaN                   NaN   \n",
       "bin3c.98     ...                   NaN                   NaN   \n",
       "\n",
       "                                              pplacer_taxonomy  \\\n",
       "user_genome                                                      \n",
       "54_67_14     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "159_36_9     d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "157_38_13    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "165_43_8     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "156_68_15    d__Bacteria;p__Proteobacteria;c__Alphaproteoba...   \n",
       "...                                                        ...   \n",
       "bin3c.78     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "bin3c.79     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.91     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.92     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "bin3c.98     d__Bacteria;p__Firmicutes_A;c__Clostridia_A;o_...   \n",
       "\n",
       "                                         classification_method  \\\n",
       "user_genome                                                      \n",
       "54_67_14     taxonomic classification defined by topology a...   \n",
       "159_36_9     taxonomic classification defined by topology a...   \n",
       "157_38_13    taxonomic classification defined by topology a...   \n",
       "165_43_8     taxonomic classification defined by topology a...   \n",
       "156_68_15    taxonomic classification defined by topology a...   \n",
       "...                                                        ...   \n",
       "bin3c.78     taxonomic classification defined by topology a...   \n",
       "bin3c.79                                                   ANI   \n",
       "bin3c.91     taxonomic classification fully defined by topo...   \n",
       "bin3c.92                taxonomic novelty determined using RED   \n",
       "bin3c.98                taxonomic novelty determined using RED   \n",
       "\n",
       "                                                          note  \\\n",
       "user_genome                                                      \n",
       "54_67_14     topological placement and ANI have congruent s...   \n",
       "159_36_9     topological placement and ANI have congruent s...   \n",
       "157_38_13    topological placement and ANI have congruent s...   \n",
       "165_43_8     topological placement and ANI have congruent s...   \n",
       "156_68_15    topological placement and ANI have congruent s...   \n",
       "...                                                        ...   \n",
       "bin3c.78                                                   NaN   \n",
       "bin3c.79                                                   NaN   \n",
       "bin3c.91                                                   NaN   \n",
       "bin3c.92                                                   NaN   \n",
       "bin3c.98                                                   NaN   \n",
       "\n",
       "            other_related_references(genome_id,species_name,radius,ANI,AF)  \\\n",
       "user_genome                                                                  \n",
       "54_67_14     GCA_005878095.1, s__13-2-20CM-66-19 sp00587809...               \n",
       "159_36_9     GCF_001881495.1, s__Rickettsiella_A isopodorum...               \n",
       "157_38_13    GCA_005882975.1, s__VBAS01 sp005882975, 95.0, ...               \n",
       "165_43_8                                                   NaN               \n",
       "156_68_15    GCF_007995055.1, s__Allosphingosinicella ginse...               \n",
       "...                                                        ...               \n",
       "bin3c.78                                                   NaN               \n",
       "bin3c.79     GCA_900540045.1, s__Firm-11 sp900540045, 95.0,...               \n",
       "bin3c.91                                                   NaN               \n",
       "bin3c.92                                                   NaN               \n",
       "bin3c.98                                                   NaN               \n",
       "\n",
       "            msa_percent  translation_table  red_value  warnings  \n",
       "user_genome                                                      \n",
       "54_67_14          95.79                 11        NaN       NaN  \n",
       "159_36_9          95.59                 11        NaN       NaN  \n",
       "157_38_13         96.72                 11        NaN       NaN  \n",
       "165_43_8          88.70                 11        NaN       NaN  \n",
       "156_68_15         97.00                 11        NaN       NaN  \n",
       "...                 ...                ...        ...       ...  \n",
       "bin3c.78          93.65                 11   0.957236       NaN  \n",
       "bin3c.79          92.30                 11   0.942301       NaN  \n",
       "bin3c.91          93.33                 11   0.864885       NaN  \n",
       "bin3c.92          94.92                 11   0.894587       NaN  \n",
       "bin3c.98          93.55                 11   0.874257       NaN  \n",
       "\n",
       "[2453 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "306767bf-2729-40d6-ae28-1fff123e18c3",
   "metadata": {},
   "source": [
    "## Dataset from Bickhart et al (695 MAGs, 286 HQ, 284 Bacteria) <a class=\"anchor\" id=\"bickhart\"></a>\n",
    "The dataset can be accessed from https://zenodo.org/record/5138306/files/hifi_das.bin3c.bins.tar.gz. The dataset are then downloaded using wget:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "543e5e9c-2430-4ffc-8091-96a17db7079b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/raw_sequences/hifi_das.bin3c.bins.tar.gz’ already there; not retrieving.\n"
     ]
    }
   ],
   "source": [
    "# MAG set for the Bickhart paper (464)\n",
    "Path(\"../data/raw_sequences\").mkdir(parents=True, exist_ok=True)\n",
    "! wget -O ../data/raw_sequences/hifi_das.bin3c.bins.tar.gz https://zenodo.org/record/5138306/files/hifi_das.bin3c.bins.tar.gz?download=1 -nc\n",
    "tar = tarfile.open(\"../data/raw_sequences/hifi_das.bin3c.bins.tar.gz\")\n",
    "tar.extractall(path=\"../data/raw_sequences/\")\n",
    "tar.close()\n",
    "#! rm ../data/raw_sequences/hifi_das.bin3c.bins.tar.gz\n",
    "\n",
    "# rename folder\n",
    "bickhart_raw_path = Path(\"../data/raw_sequences/b3c_hifi_dastool/\")\n",
    "try:\n",
    "    bickhart_raw_path.rename(\"../data/raw_sequences/Bickhart_et_al/\")\n",
    "except FileExistsError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5b154a-8f4d-4f48-9965-0ad2fa766d97",
   "metadata": {},
   "source": [
    "The metadata for the genomes are then cleaned to create a list of genomes that will be used in the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3bc1bd17-82dd-4fc0-86c9-a8e323e3da9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/supplementary_tables/41587_2021_1130_MOESM3_ESM.xlsx’ already there; not retrieving.\n",
      "\n",
      "Bickhart et al has a total of 695 bins. Based on their methods, 353 are HQ.\n",
      "After running CheckM on the 695 bins, we get 286 HQ MAGs.\n",
      "These bins are classified as HQ in CheckM, but not in DAStools:\n",
      "['bin3c.119', 'bin3c.136', 'bin3c.169', 'bin3c.172', 'bin3c.181', 'bin3c.182', 'bin3c.183', 'bin3c.191', 'bin3c.212', 'bin3c.214', 'bin3c.219', 'bin3c.232', 'bin3c.239', 'bin3c.252', 'bin3c.254', 'bin3c.26', 'bin3c.266', 'bin3c.274', 'bin3c.276', 'bin3c.292', 'bin3c.345', 'bin3c.347', 'bin3c.396', 'bin3c.400', 'bin3c.42', 'bin3c.426', 'bin3c.460', 'bin3c.465', 'bin3c.484', 'bin3c.490', 'bin3c.497', 'bin3c.543', 'bin3c.623', 'bin3c.74', 'bin3c.90', 'bin3c.94']\n",
      "\n",
      "Filtering for bacteria only genomes...\n",
      "Genome: 'bin3c.221' is not in GTDB file\n",
      "Genome: 'bin3c.602' is not in GTDB file\n",
      "Found 284/286 Bacterial Genomes\n"
     ]
    }
   ],
   "source": [
    "# Download supplementary materials from the paper\n",
    "! wget -P ../data/supplementary_tables/ https://static-content.springer.com/esm/art%3A10.1038%2Fs41587-021-01130-z/MediaObjects/41587_2021_1130_MOESM3_ESM.xlsx -nc\n",
    "\n",
    "# EDA - How many HQ MAGs based on DAStool definition?\n",
    "df_bickhart = pd.read_excel(\"../data/supplementary_tables/41587_2021_1130_MOESM3_ESM.xlsx\", 'SupplementaryTable2')\n",
    "df_bickhart.loc[:, \"genome_id\"] = [i.replace(\"_\", \".\") for i in df_bickhart.loc[:, 'HiFiBin']]\n",
    "df_bickhart = df_bickhart.set_index(\"genome_id\", drop=False)\n",
    "complete = df_bickhart.loc[:, \"HiFiBin Completeness\"] > 90\n",
    "contam = df_bickhart.loc[:, \"HifiBin Contamination\"] < 5\n",
    "df_bickhart_HQ = df_bickhart.loc[complete & contam]\n",
    "print(f\"Bickhart et al has a total of {len(df_bickhart)} bins. Based on their methods, {len(df_bickhart_HQ)} are HQ.\")\n",
    "\n",
    "# Caitlin has ran CheckM to all bins. How many HQ MAGs based on CheckM definition?\n",
    "df_bickhart_checkm = pd.read_csv(file_bickhart_checkm, sep=\"\\t\")\n",
    "df_bickhart_checkm.loc[:, \"genome_id\"] = [i.replace(\".contigs\", \"\") for i in df_bickhart_checkm.loc[:, \"Bin Id\"]]\n",
    "df_bickhart_checkm = df_bickhart_checkm.set_index(\"genome_id\", drop=False)\n",
    "complete = df_bickhart_checkm.loc[:, \"Completeness\"] > 90\n",
    "contam = df_bickhart_checkm.loc[:, \"Contamination\"] < 5\n",
    "df_bickhart_checkm_HQ = df_bickhart_checkm.loc[complete & contam]\n",
    "print(f\"After running CheckM on the {len(df_bickhart_checkm)} bins, we get {len(df_bickhart_checkm_HQ)} HQ MAGs.\")\n",
    "\n",
    "# Where does the tools disagree?\n",
    "try:\n",
    "    df_bickhart_HQ.loc[list(df_bickhart_checkm_HQ.index)]\n",
    "except KeyError as e:\n",
    "    print(\"These bins are classified as HQ in CheckM, but not in DAStools:\")\n",
    "    print(ast.literal_eval(e.args[0].replace(\" not in index\", \"\")))\n",
    "    \n",
    "# Filter out archaeal genomes\n",
    "print(\"\\nFiltering for bacteria only genomes...\")\n",
    "filter_mask = []\n",
    "for i in df_bickhart_checkm_HQ.index:\n",
    "    try:\n",
    "        tax = df_gtdb.loc[i, \"classification\"]\n",
    "        if tax.startswith(\"d__Bacteria\"):\n",
    "            filter_mask.append(i)\n",
    "        else:\n",
    "            print(f\"Genome: {i} is not bacteria.\")\n",
    "            print(tax)\n",
    "    except KeyError as e:\n",
    "        print(f\"Genome: {e} is not in GTDB file\")\n",
    "print(f\"Found {len(filter_mask)}/{len(df_bickhart_checkm_HQ)} Bacterial Genomes\")\n",
    "df_bickhart_checkm_HQ = df_bickhart_checkm_HQ.loc[filter_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c6566a3-d869-4090-adae-33cecc6c5478",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format to BGCflow config\n",
    "df_bickhart = pd.concat([df_bgcflow_template, df_bickhart_checkm_HQ])\n",
    "col = df_bickhart.pop(\"genome_id\")\n",
    "df_bickhart.insert(0, col.name, col)\n",
    "df_bickhart.loc[:, \"source\"] = \"custom\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "257e4c53-286c-4019-9545-e18f4f0a5e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bickhart.to_csv(external_table_path / \"df_bickhart_checkm_HQ.csv\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de0e9fb7-0d05-4fec-a944-48bef9e9c0ee",
   "metadata": {},
   "source": [
    "The selected genomes are then copied to the BGCflow directory for downstream analyses:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3090a265-6f6c-4a0b-8769-cc236b2135d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Up Paths\n",
    "bickhart_raw_path = Path(\"../data/raw_sequences/Bickhart_et_al/\")\n",
    "bickhart_data_path = external_data_path / \"Bickhart_et_al\"\n",
    "bickhart_data_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# copy files to target dir\n",
    "for i in df_bickhart.loc[:, \"genome_id\"]:\n",
    "    item = bickhart_raw_path / f\"flye4.das_DASTool_bins/{i}.contigs.fa.gz\"\n",
    "    dest = bickhart_data_path / f\"{i}.fna\"\n",
    "    if dest.is_file():\n",
    "        pass\n",
    "    else:\n",
    "        if item.is_file():\n",
    "            with gzip.open(item, 'rb') as f_in:\n",
    "                with open(dest, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)\n",
    "        else:\n",
    "            item = bickhart_raw_path / f\"flye4.das_DASTool_bins/{i}.contigs.fa\"\n",
    "            assert(item.is_file())\n",
    "            shutil.copy(item, dest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7af788-5f23-4e77-82b4-7819f6a35c23",
   "metadata": {},
   "source": [
    "# Dataset from Liu et al (557 MAGs, 153 HQ, 153 Bacteria) <a class=\"anchor\" id=\"liu\"></a>\n",
    "The dataset was kindly provided by the authors here: https://www.dropbox.com/sh/qj4aginbflqmxhq/AAB5INvLNKrGiBtjpRbteYJ7a. The dataset are then downloaded using wget:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "469dfd96-7d1c-49e0-897d-6e4910540f0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/raw_sequences/Liu_et_al.zip’ already there; not retrieving.\n",
      "mkdir: cannot create directory ‘../data/raw_sequences/Liu_et_al’: File exists\n",
      "warning:  stripped absolute path spec from /\n",
      "mapname:  conversion of  failed\n"
     ]
    }
   ],
   "source": [
    "# MAG set for the Liu paper (557)\n",
    "! wget -O ../data/raw_sequences/Liu_et_al.zip https://www.dropbox.com/sh/qj4aginbflqmxhq/AAB5INvLNKrGiBtjpRbteYJ7a?dl=0 -nc\n",
    "! mkdir ../data/raw_sequences/Liu_et_al\n",
    "! unzip -nq ../data/raw_sequences/Liu_et_al.zip  -d ../data/raw_sequences/Liu_et_al"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10b5943-91b7-46e9-ae6a-56dcc777c538",
   "metadata": {},
   "source": [
    "The metadata for the genomes are then cleaned to create a list of genomes that will be used in the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c526a32c-49cc-4ace-b3b4-dc3ced0c59b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/supplementary_tables/40168_2021_1155_MOESM2_ESM.xlsx’ already there; not retrieving.\n",
      "\n",
      "Liu et al has 557 MAGs.\n",
      "Out of these, 153 is HQ\n"
     ]
    }
   ],
   "source": [
    "# Download supplementary materials from the paper\n",
    "! wget -P ../data/supplementary_tables/ https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-021-01155-1/MediaObjects/40168_2021_1155_MOESM2_ESM.xlsx -nc\n",
    "\n",
    "# Clean Up\n",
    "df_liu = pd.read_excel(\"../data/supplementary_tables/40168_2021_1155_MOESM2_ESM.xlsx\", sheet_name=2, skiprows=0)\n",
    "print(f\"Liu et al has {len(df_liu)} MAGs.\")\n",
    "\n",
    "# Filter for HQ only\n",
    "complete = df_liu.loc[:, \"Completeness (%)\"] > 90\n",
    "contam = df_liu.loc[:, \"Contamination (%)\"] < 5\n",
    "df_liu_HQ = df_liu.loc[complete & contam]\n",
    "print(f\"Out of these, {len(df_liu_HQ)} is HQ\")\n",
    "\n",
    "df_liu_HQ = pd.concat([df_bgcflow_template, df_liu_HQ])\n",
    "df_liu_HQ.insert(0, 'genome_id', df_liu_HQ.loc[:, \"MAGs\"])\n",
    "df_liu_HQ = df_liu_HQ.drop(columns=[\"MAGs\"])\n",
    "df_liu_HQ.loc[:, \"source\"] = \"custom\"\n",
    "df_liu_HQ.to_csv(external_table_path / \"df_liu_checkm_HQ.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0b5c18ad-3cba-432e-9156-ebb33b96a8e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Up Paths\n",
    "liu_raw_path = Path(\"../data/raw_sequences/Liu_et_al/\")\n",
    "liu_data_path = external_data_path / \"Liu_et_al\"\n",
    "liu_data_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# copy files to target dir\n",
    "for i in df_liu_HQ.loc[:, \"genome_id\"]:\n",
    "    item = liu_raw_path / f\"{i}.fasta\"\n",
    "    dest = liu_data_path / f\"{i}.fna\"\n",
    "    if dest.is_file():\n",
    "        pass\n",
    "    else:\n",
    "        if item.is_file():\n",
    "            with open(item, 'rb') as f_in:\n",
    "                with open(dest, 'wb') as f_out:\n",
    "                    shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b89e3cd-a8fb-49d8-bd34-d1f24b08153d",
   "metadata": {},
   "source": [
    "# Dataset from Christoph et al (376 MAGs, 73 HQ, 73 Bacteria) <a class=\"anchor\" id=\"christoph\"></a>\n",
    "The dataset from Christoph et al are publicly available from NCBI: https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA449266. The assembly details file: `PRJNA449266_AssemblyDetails.txt` were downloaded, containing the assembly accessions of the study, which then can be downloaded using ncbi-genome-download.\n",
    "\n",
    "HQ-MAGs are selected from the dataset with this definition:\n",
    ">High-quality draft' will indicate that a SAG or MAG is >90% complete with less than 5% contamination\n",
    "\n",
    "The metadata for the genomes are then cleaned to create a list of genomes that will be used in the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e847f352-f9af-447d-ab48-8f9085b05845",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/supplementary_tables/41586_2018_207_MOESM3_ESM.xlsx’ already there; not retrieving.\n",
      "\n",
      "Christoph et al has: 376 MAGs\n",
      "Out of these, 73 are HQ MAGs\n"
     ]
    }
   ],
   "source": [
    "# Download supplementary materials from the paper\n",
    "! wget -P ../data/supplementary_tables/ https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-018-0207-y/MediaObjects/41586_2018_207_MOESM3_ESM.xlsx -nc\n",
    "df_christoph = pd.read_excel(\"../data/supplementary_tables/41586_2018_207_MOESM3_ESM.xlsx\", sheet_name=0, skiprows=0)\n",
    "print(f\"Christoph et al has: {len(df_christoph)} MAGs\")\n",
    "\n",
    "# filter for HQ\n",
    "a = df_christoph.loc[:, \"CheckM Completeness %\"] > 90\n",
    "b = df_christoph.loc[:, \"CheckM Contamination %\"] < 5\n",
    "df_christoph = df_christoph[a & b]\n",
    "print(f\"Out of these, {len(df_christoph)} are HQ MAGs\")\n",
    "\n",
    "# Get NCBI accession\n",
    "df_christoph.loc[:, \"ncbi_mapping\"] = [i.split(\"_\", 1)[-1].replace(\"_\", \" \") for i in df_christoph.loc[:, \"Genome\"]]\n",
    "need_cleaning = df_christoph.loc[:, \"ncbi_mapping\"].str.startswith(\"unk\")\n",
    "df_christoph.loc[df_christoph[need_cleaning].index, \"ncbi_mapping\"] = [i.split(\" \")[-1] for i in df_christoph[need_cleaning].loc[:, \"ncbi_mapping\"]]\n",
    "df_christoph_ncbi = pd.read_csv(file_christoph_ncbi, sep=\"\\t\", skiprows=1, index_col=False)\n",
    "df_christoph_ncbi_subset = df_christoph_ncbi[df_christoph_ncbi.loc[:, \"Isolate\"].isin(df_christoph.loc[:, \"ncbi_mapping\"])]\n",
    "df_christoph = df_christoph.merge(df_christoph_ncbi_subset, how=\"left\", left_on=\"ncbi_mapping\", right_on=\"Isolate\")\n",
    "df_christoph = df_christoph.set_index(\"# Assembly\", drop=False)\n",
    "\n",
    "# Format to BGCflow config\n",
    "df_christoph = pd.concat([df_bgcflow_template, df_christoph])\n",
    "df_christoph.insert(0, 'genome_id', df_christoph.loc[:, \"# Assembly\"])\n",
    "df_christoph = df_christoph.drop(columns=[\"# Assembly\"])\n",
    "df_christoph.loc[:, \"source\"] = \"ncbi\"\n",
    "df_christoph.to_csv(external_table_path / \"df_christoph_checkm_HQ.csv\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "292195b2-26c4-4ae5-b891-36950cf9b371",
   "metadata": {},
   "source": [
    "PS: Some of these genomes are already used as GTDB reference. These genomes are discarded in the GTDB-tk classify wf as it will return an error:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "05ec5bec-c3d9-4a78-88ea-554f47a7a272",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv(file_gtdb_warning, skiprows=1, header=None, sep=\" \").loc[:, 3].to_csv(\"../tables/filter_gtdb.txt\", header=None, index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1c757e-5833-4a7b-bc1f-723e29b50f6d",
   "metadata": {},
   "source": [
    "# Dataset from Sharrar et al (1334 MAGs, 374 HQ, ) <a class=\"anchor\" id=\"sharrar\"></a>\n",
    "\n",
    "The dataset was kindly provided by the authors here: https://figshare.com/ndownloader/files/18105260. The dataset are then downloaded using wget:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "250b1f0b-99f6-4bc4-9dc6-bd348a831670",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/raw_sequences/1334_genomes.tar.gz’ already there; not retrieving.\n"
     ]
    }
   ],
   "source": [
    "! wget -O ../data/raw_sequences/1334_genomes.tar.gz https://figshare.com/ndownloader/files/18105260 -nc\n",
    "tar = tarfile.open(\"../data/raw_sequences/1334_genomes.tar.gz\")\n",
    "tar.extractall(path=\"../data/raw_sequences/\")\n",
    "tar.close()\n",
    "\n",
    "# rename folder\n",
    "sharrar_raw_path = Path(\"../data/raw_sequences/1334_genomes/\")\n",
    "try:\n",
    "    sharrar_raw_path.rename(\"../data/raw_sequences/Sharrar_et_al/\")\n",
    "except FileExistsError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9d36e9-6e17-4de8-95f9-67f8da29ac45",
   "metadata": {},
   "source": [
    "The metadata for the genomes are then cleaned to create a list of genomes that will be used in the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c011def3-9207-4ae4-9f11-feb50a6a27e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘../data/supplementary_tables/mbio.00416-20-st002.xlsx’ already there; not retrieving.\n",
      "\n",
      "Sharrar et al has 1334 MAGs.\n",
      "Out of these, 374 are HQ.\n"
     ]
    }
   ],
   "source": [
    "# Download supplementary materials from the paper\n",
    "! wget -P ../data/supplementary_tables/ https://journals.asm.org/doi/suppl/10.1128/mBio.00416-20/suppl_file/mbio.00416-20-st002.xlsx -nc\n",
    "\n",
    "# Clean Up\n",
    "df_sharrar = pd.read_excel(\"../data/supplementary_tables/mbio.00416-20-st002.xlsx\", sheet_name=0, skiprows=0)\n",
    "print(f\"Sharrar et al has {len(df_sharrar)} MAGs.\")\n",
    "df_sharrar = df_sharrar[df_sharrar.loc[:, \"MAG draft quality\"] == \"High\"]\n",
    "print(f\"Out of these, {len(df_sharrar)} are HQ.\")\n",
    "\n",
    "# Format to BGCflow\n",
    "df_sharrar = pd.concat([df_bgcflow_template, df_sharrar])\n",
    "df_sharrar.insert(0, 'genome_id', df_sharrar.loc[:, \"Genome name\"])\n",
    "df_sharrar = df_sharrar.drop(columns=[\"Genome name\"])\n",
    "df_sharrar.loc[:, \"source\"] = \"custom\"\n",
    "df_sharrar.to_csv(external_table_path / \"df_sharrar.csv\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec25065b-70c8-4cb1-8845-c681a5a66544",
   "metadata": {},
   "source": [
    "The selected genomes are then copied to the BGCflow directory for downstream analyses:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3e65bc22-df39-4291-9819-c1ae35286f5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy files to target dir\n",
    "sharrar_raw_path = Path(\"../data/raw_sequences/Sharrar_et_al\")\n",
    "sharrar_data_path = external_data_path / \"Sharrar_et_al\"\n",
    "sharrar_data_path.mkdir(parents=True, exist_ok=True)\n",
    "for i in df_sharrar.loc[:, \"genome_id\"]:\n",
    "    item = sharrar_raw_path / f\"{i}.contigs.fa\"\n",
    "    dest = sharrar_data_path / f\"{i}.fna\"\n",
    "    shutil.copy(item, dest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28e7bb91-28dd-42c7-a678-0f0811a513cb",
   "metadata": {},
   "source": [
    "### Sanitize Genome and Contig Names\n",
    "Apparently, the contig ids from Chen et al needs to be sanitized because of its length."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "215ec77b-6ab8-4f25-8ff8-be9d8254591a",
   "metadata": {},
   "source": [
    "Here, we will attempt to shorten the ids by extracting unique names from the original ids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "06ad172e-c7b9-43cf-9ae5-ee4d710fc0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Grabbing all ids from .fna files with glob...\n",
      "There are 7 different variation of ids, each with N number of keywords:\n",
      "{2, 3, 4, 8, 9, 10, 14}\n",
      "Manually curating each id types...\n",
      "Recording id changes as dict...\n",
      "Correcting contig names based on dictionary..\n",
      "Changes are recorded as json files..\n",
      "Sanitized genomes are saved in ../../bgcflow/data/external/Sharrar_et_al_sanitized\n"
     ]
    }
   ],
   "source": [
    "# Prepare path for correction\n",
    "sharrar_path = external_data_path / \"Sharrar_et_al/\"\n",
    "sharrar_corrected_path = external_data_path / \"Sharrar_et_al_sanitized/\"\n",
    "sharrar_corrected_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "print(\"Grabbing all ids from .fna files with glob...\")\n",
    "sample_ids = [i.stem for i in sharrar_path.glob(\"*.fna\")]\n",
    "\n",
    "# how many id variables are there?\n",
    "id_variables = set([len(i.split(\"_\")) for i in sample_ids])\n",
    "print(f\"There are {len(id_variables)} different variation of ids, each with N number of keywords:\")\n",
    "print(id_variables)\n",
    "\n",
    "print(\"Manually curating each id types...\")\n",
    "# get unique values for each id types\n",
    "type_2 = {i:i.split(\"_\")[-1] for i in sample_ids if len(i.split(\"_\")) == 2}\n",
    "type_3 = {i:i.split(\"_\", 1)[-1] for i in sample_ids if len(i.split(\"_\")) == 3}\n",
    "type_4 = {i:i.split(\"_\", 2)[-1] for i in sample_ids if len(i.split(\"_\")) == 4}\n",
    "type_8 = {i:i.split(\"_\", 5)[-1] for i in sample_ids if len(i.split(\"_\")) == 8}\n",
    "type_9 = {i:i.split(\"_\", 6)[-1] for i in sample_ids if len(i.split(\"_\")) == 9}\n",
    "type_10 = {i:i.split(\"_\", 7)[-1] for i in sample_ids if len(i.split(\"_\")) == 10}\n",
    "type_14 = {i:i.split(\"_\", 11)[-1] for i in sample_ids if len(i.split(\"_\")) == 14}\n",
    "\n",
    "print(\"Recording id changes as dict...\")\n",
    "# append into one dictionary\n",
    "sample_id_dicts = {}\n",
    "unique_ids = [type_2, type_3, type_4, type_8, type_9, type_10, type_14]\n",
    "for d in unique_ids:\n",
    "    sample_id_dicts.update(d)\n",
    "\n",
    "print(\"Correcting contig names based on dictionary..\")\n",
    "print(\"Changes are recorded as json files..\")\n",
    "# Correcting contig names based on dicts\n",
    "output_dir = sharrar_corrected_path\n",
    "input_files = [i for i in sharrar_path.glob(\"*.fna\")]\n",
    "\n",
    "for i in input_files:\n",
    "    original_file = Path(i)\n",
    "    new_genome_id = sample_id_dicts[i.stem]\n",
    "    corrected_file = Path(output_dir) / f\"{new_genome_id}.fna\"\n",
    "    \n",
    "    log_dict = {}\n",
    "    \n",
    "    # Use biopython to change contig names\n",
    "    with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "        records = SeqIO.parse(original_file, 'fasta')\n",
    "        ctr = 1\n",
    "        for record in records:       \n",
    "            old_id = record.id\n",
    "            new_id = f\"{new_genome_id}_{ctr}\"\n",
    "            record.id = new_id\n",
    "            ctr = ctr + 1\n",
    "            SeqIO.write(record, corrected, 'fasta')\n",
    "            \n",
    "            log_dict[old_id] = new_id\n",
    "    \n",
    "    with open(Path(output_dir) / f\"{new_genome_id}.json\", \"w\") as out_file:\n",
    "        json.dump(log_dict, out_file, indent = 4)\n",
    "    \n",
    "print(f\"Sanitized genomes are saved in {sharrar_corrected_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "82336722-19bb-4ed7-9dbb-22a584199dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update config files to the new ids\n",
    "df_sharrar = pd.read_csv(external_table_path / \"df_sharrar.csv\")\n",
    "df_sharrar = df_sharrar.rename(columns={\"genome_id\":\"original_genome_id\"})\n",
    "for i in df_sharrar.index:\n",
    "    df_sharrar.loc[i, \"genome_id\"] = sample_id_dicts[df_sharrar.loc[i, \"original_genome_id\"]]\n",
    "    \n",
    "# shift column 'Name' to first position\n",
    "first_column = df_sharrar.pop('genome_id')\n",
    "  \n",
    "# insert column using insert(position,column_name,first_column) function\n",
    "df_sharrar.insert(0, 'genome_id', first_column)\n",
    "df_sharrar = df_sharrar.set_index(\"genome_id\", drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f616d1ab-f55c-4456-9d1f-8d20283449e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of the HQ MAGs, 350 are bacteria\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    df_gtdb.loc[df_sharrar.index]\n",
    "except KeyError as e:\n",
    "    err = ast.literal_eval(e.args[0].replace(\" not in index\", \"\"))\n",
    "    df_sharrar.loc[:, \"classification\"] = df_gtdb.loc[[i for i in df_sharrar.index if i not in err]].classification\n",
    "    df_sharrar = df_sharrar.drop(err)\n",
    "print(f\"Out of the HQ MAGs, {len(df_sharrar)} are bacteria\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dff00684-2a7a-430d-8bb4-6d022488b9f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sharrar.to_csv(external_table_path / \"df_sharrar_checkm_HQ.csv\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9af38a4d-5230-457a-95e9-434869c5f5d4",
   "metadata": {},
   "source": [
    "# Dataset from Chen et al (96 MAGs, 24 HQ, 21 Bacteria) <a class=\"anchor\" id=\"chen\"></a>\n",
    "The dataset for Chen et al are available from: https://api.mg-rast.org/project/mgp81948?verbosity=full. The genomes can then be fetched using MG-RAST API:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75f3d0f0-ad87-4f69-97df-3c1f75158e78",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_api(api_url, output):\n",
    "    response = requests.get(api_url)\n",
    "    project_metadata = response.json()\n",
    "    with open(output, 'w') as outfile:\n",
    "        json.dump(project_metadata, outfile)\n",
    "    return None\n",
    "\n",
    "get_api('https://api.mg-rast.org/project/mgp81948?verbosity=full', '../data/raw_sequences/mgp81948.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5d62385c-4072-4422-85a4-77edcadd77f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"../data/raw_sequences/Chen_et_al\").mkdir(parents=True, exist_ok=True)\n",
    "with open('../data/raw_sequences/mgp81948.json', \"r\") as f:\n",
    "    study_metadata = json.load(f)\n",
    "    ids = [i[\"metagenome_id\"] for i in study_metadata[\"metagenomes\"]]\n",
    "    for mg_id in ids:\n",
    "        file = Path(f\"../data/raw_sequences/Chen_et_al/{mg_id}.fna\")\n",
    "        if file.is_file():\n",
    "            pass\n",
    "        else:\n",
    "            url = f\"https://api.mg-rast.org/download/{mg_id}?file=299.1\"\n",
    "            ! wget -O ../data/raw_sequences/Chen_et_al/{mg_id}.fna {url} -nc #2>> ../data/raw_sequences/Chen_et_al/download.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b031dae-e443-4dd5-a059-0e3b349ac98e",
   "metadata": {},
   "source": [
    "The metadata for the genomes can be accessed from https://www.frontiersin.org/articles/10.3389/fmicb.2020.01950/full#supplementary-material. The metadata are then cleaned to create a list of genomes that will be used in the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "aefc6748-4753-4b4d-b4c1-39d818096b40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chen et al has 96 MAGs\n",
      "Out of these, 24 are HQ MAGs\n",
      "Out of the HQ MAGs, 21 are bacteria\n"
     ]
    }
   ],
   "source": [
    "# Download supplementary materials from the paper\n",
    "df_chen_metadata = pd.read_excel(file_chen_metadata, sheet_name=0, skiprows=5)\n",
    "\n",
    "with open(\"../data/raw_sequences/mgp81948.json\", \"r\") as file:\n",
    "    chen_metadata = json.load(file)\n",
    "df_chen = pd.DataFrame.from_dict([i for i in chen_metadata[\"metagenomes\"]]).set_index(\"metagenome_id\", drop=False)\n",
    "print(f\"Chen et al has {len(df_chen)} MAGs\")\n",
    "\n",
    "df_chen_checkm = pd.read_csv(file_chen_checkm).set_index(\"genome_id\")\n",
    "df_chen = pd.merge(df_chen, df_chen_checkm, left_index=True, right_index=True)\n",
    "df_chen.columns\n",
    "\n",
    "# filter for HQ\n",
    "a = df_chen.loc[:, \"Completeness\"] > 90\n",
    "b = df_chen.loc[:, \"Contamination\"] < 5\n",
    "df_chen = df_chen[a & b]\n",
    "print(f\"Out of these, {len(df_chen)} are HQ MAGs\")\n",
    "\n",
    "# filter archaea\n",
    "try:\n",
    "    df_chen.loc[:, \"classification\"] = df_gtdb.loc[df_chen.index].classification\n",
    "except KeyError as e:\n",
    "    err = ast.literal_eval(e.args[0].replace(\" not in index\", \"\"))\n",
    "    df_chen.loc[:, \"classification\"] = df_gtdb.loc[[i for i in df_chen.index if i not in err]].classification\n",
    "    df_chen = df_chen.drop(err)\n",
    "print(f\"Out of the HQ MAGs, {len(df_chen)} are bacteria\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1e8ffb69-a121-475c-aa81-c53db57eb8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format to BGCflow\n",
    "df_chen = pd.concat([df_bgcflow_template, df_chen])\n",
    "df_chen.insert(0, 'genome_id', df_chen.loc[:, \"metagenome_id\"])\n",
    "df_chen = df_chen.drop(columns=[\"metagenome_id\"])\n",
    "df_chen.loc[:, \"source\"] = \"custom\"\n",
    "df_chen.to_csv(external_table_path / \"df_chen_checkm_HQ.csv\", index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5930a7b-6f72-453b-af0c-38a77fac6f28",
   "metadata": {},
   "source": [
    "## Sanitize Contig Names\n",
    "Apparently, the contig names Chen et al needs to be sanitized because of its length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d6e5521c-b8eb-4d25-8e71-9bd62b68cf4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sanitize_chen_et_al(input_file, output_dir):\n",
    "    \"\"\"\n",
    "    Shorten contig name\n",
    "    \"\"\"\n",
    "    original_file = Path(input_file)\n",
    "    corrected_file = Path(output_dir) / f\"{original_file.stem}.fna\"\n",
    "    \n",
    "    log_dict = {}\n",
    "\n",
    "    with open(original_file) as original, open(corrected_file, 'w') as corrected:\n",
    "        records = SeqIO.parse(original_file, 'fasta')\n",
    "        ctr = 1\n",
    "        for record in records:       \n",
    "            old_id = record.id\n",
    "            new_id = f\"{original_file.stem}_{ctr}\"\n",
    "            record.id = new_id\n",
    "            ctr = ctr + 1\n",
    "            SeqIO.write(record, corrected, 'fasta')\n",
    "            \n",
    "            log_dict[old_id] = new_id\n",
    "    \n",
    "    with open(Path(output_dir) / f\"{original_file.stem}.json\", \"w\") as out_file:\n",
    "        json.dump(log_dict, out_file, indent = 4)\n",
    "    return\n",
    "\n",
    "chen_path = Path(\"../data/raw_sequences/Chen_et_al\")\n",
    "chen_corrected_path = external_data_path / \"Chen_et_al_sanitized\"\n",
    "chen_corrected_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "for i in chen_path.glob(\"*.fna\"):\n",
    "    sanitize_chen_et_al(i, chen_corrected_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c8ee67e-d5d9-428e-95c8-f2add2e626f4",
   "metadata": {},
   "source": [
    "# Merge all tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a28c1474-e459-489f-8eff-bf20a7dea7f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_bickhart_checkm_HQ.csv  df_christoph_checkm_HQ.csv  df_sharrar.csv\n",
      "df_chen_checkm_HQ.csv\t   df_liu_checkm_HQ.csv        df_sharrar_checkm_HQ.csv\n"
     ]
    }
   ],
   "source": [
    "! ls ../tables/other_studies_HQ/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1661f2c7-62ad-4722-8b82-433b854178fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_HQ = [pd.read_csv(i) for i in Path(\"../tables/other_studies_HQ/\").glob(\"*HQ*\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e39e423b-599e-4e6d-b008-9c0a0e3e96fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_singleton = pd.read_csv(\"../../bgcflow/config/Singleton_et_al/df_singleton_bacteria.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9c87271f-4c54-4c37-9c9e-7f76fd6109a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_HQ.append(df_singleton)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1df40393-026d-43a3-9fee-092e623ac026",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_study = pd.concat(dfs_HQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3562b480-4f26-4ae3-8041-5c14af1930f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"../../bgcflow/config/all_hqmags_studies\").mkdir(parents=True, exist_ok=True)\n",
    "df_all_study.to_csv(\"../../bgcflow/config/all_hqmags_studies/all_hqmags_studies_checkm_HQ.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
