```{r}
library(tidyverse)
library(plotly)
library(treeio)
library(ggtree)
library(ggtreeExtra)
library(svglite)
library(jsonlite)
library(rvest)
```

Fetch info for Nitrospirota genomes

```{r}
bioproject_accession <- "prjna629478"

assembly_details <- 
  paste0("https://www.ncbi.nlm.nih.gov/bioproject/", bioproject_accession) %>%
  read_html() %>%
  html_node("#AssemblyDetails") %>%
  html_table() %>% 
  select(Assembly, Isolate) %>%
setNames(c("genome_id", "isolate"))

gtdb202_taxonomy <- read_tsv(
  "data/gtdbtk.bac120.summary.tsv", 
  col_select = 1:2, 
  col_types = "cc") %>%
  set_names(c("genome_id", "gtdb_taxonomy"))

singleton_2021_table3_url <-
  "https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-021-22203-2/MediaObjects/41467_2021_22203_MOESM5_ESM.xlsx"

download.file(singleton_2021_table3_url,
              destfile = "data/singleton_2021_table3.xlsx", 
              method = "curl")

gtdb_metadata_url <- 
  "https://data.gtdb.ecogenomic.org/releases/release202/202.0/bac120_metadata_r202.tar.gz"
tar_file <- "data/bac120_metadata_r202.tar.gz"
download.file(gtdb_metadata_url, tar_file)
untar(tar_file, exdir = 'data')

mags <- read_xlsx(
  "data/singleton_2021_table3.xlsx",
  skip = 1) %>%
  mutate(isolate = str_replace(MAG, '.fa', ""), 
         assembly_level = if_else(Circ == "CIRCULAR",
                                  "Complete/Chromosome",
                                  "Contig/Scaffold"),
         checkm_completeness = Comp, 
         checkm_contamination = Cont, 
         genome_size = TotBP, 
         ncbi_bioproject = "PRJNA629478",
         mimag_quality = "HQ",
         source = "genbank",
         .keep = "none") %>%
  left_join(assembly_details, by = "isolate") %>%
  left_join(gtdb202_taxonomy, by = "genome_id") %>%
  select(-isolate) %>%
  filter(str_detect(gtdb_taxonomy, "p__Nitrospirota;")) 

metadata <- read_tsv("data/bac120_metadata_r202.tsv", 
                    col_select = c(1, 3:4, 14, 17, 41:43, 46, 49)) %>%
  filter(str_detect(gtdb_taxonomy, "p__Nitrospirota;")) %>%
  mutate(genome_id = str_sub(accession, start = 4), 
         source = if_else(str_detect(accession, "^RS_"), "refseq", "genbank"), 
         mimag_quality = case_when(
           mimag_high_quality ~ "HQ", 
           mimag_medium_quality ~ "MQ", 
           mimag_low_quality ~ "LQ"), 
         assembly_level = case_when(
           ncbi_assembly_level == "Contig" ~ "Contig/Scaffold", 
           ncbi_assembly_level == "Scaffold" ~ "Contig/Scaffold", 
           ncbi_assembly_level == "Complete Genome" ~ "Complete/Chromosome",
           ncbi_assembly_level == "Chromosome" ~ "Complete/Chromosome"),
         .keep = "unused") %>%
  filter(mimag_quality != "LQ") 

genomes <- bind_rows(mags, metadata) 

genomes
```

Parse info into tables

```{r include = FALSE}

bgcs <- read_csv("tables/p__Nitrospirota_all_bgcs.csv", 
                 show_col_types = FALSE) %>%
  filter(genome_id %in% genomes$genome_id)

gtdb_tree <- read.newick("data/gtdbtk.bac120.classify.tree")

taxon <- "p__Nitrospirota"

taxon_node <- gtdb_tree$node.label[str_detect(gtdb_tree$node.label, taxon)][1]

phylum_tree <- tree_subset(gtdb_tree, node = taxon_node, levels_back = 0)

phylum_tree$node.label <- str_extract(phylum_tree$node.label, '.__.*$') %>%
  str_replace_all('\'', '')

genomes <- genomes %>%
  left_join(
    bgcs %>% 
      group_by(genome_id) %>% 
      summarise(total_bgcs = n(),
                bgcs_on_contig_edge = sum(contig_edge))) %>%
  mutate(across(contains("bgcs"), ~replace_na(.x, 0))) %>%
  relocate(genome_id, contains('bgcs'), gtdb_taxonomy)
  
write_csv(bgcs, "tables/p__Nitrospirota_all_bgcs.csv")
write_csv(genomes, "tables/p__Nitrospirota_all_genomes.csv")
```

Customize BiG-SCAPE classes and color palette

```{r}
bigscape_classes <- c("Terpene",
                      "RiPPs", 
                      "NRPS",
                      "PKS-NRP_Hybrids",
                      "PKSI",
                      "PKSother", 
                      "Saccharides",
                      "Others")                             

my_classes <- c("Terpene", 
                "RiPP", 
                "NRPS", 
                "NRPS-PKS", 
                "PKS", 
                "PKS", 
                "Other", 
                "Other")

my_colors <- c("#ff8800",
               "#20c200",
               "#ff0000",
               "#dd00ff",
               "#0d00ff",
               "#757575")
```

QC Boxplots for the phylum Nitrospirota

```{r, echo = FALSE}

#Create dataset

boxplots_ds <- genomes %>%
 
  mutate(perc_contig_edge = bgcs_on_contig_edge/total_bgcs, 
         quality = case_when(
           ncbi_bioproject == "PRJNA629478" ~ "AS hqMAG", 
           assembly_level == "Complete/Chromosome" ~ assembly_level,  
           assembly_level == "Contig/Scaffold" ~ mimag_quality), 
         taxonomy = str_replace_all(gtdb_taxonomy, ";", "\n"))

#Create plot

boxplots <- ggplot(boxplots_ds, 
       aes(x = quality, 
           y = perc_contig_edge, 
           label = taxonomy)) +
  geom_boxplot() +
  geom_jitter(aes(color = source), 
              alpha = 0.5,
              height = 0.02) +
  scale_color_manual(values = c('red', 'black')) +
  theme_minimal() +
  theme(axis.text.x = element_text(hjust = 0.5))+
  ylab("% of BGCs on a contig edge") +
  xlab("")

#Save figure

ggsave(boxplots, 
       filename = "figures/boxplot_myxo.svg", 
       device = "svg", 
       height = 4, 
       width = 4.5, 
       units = "in")

#Interactive figure

ggplotly(boxplots)

```

Overview of the phylum Nitrospirota Add number of MAGs per clade

```{r echo=FALSE, warning = FALSE, fig.height=10, fig.width=7}

phylum_tree_ds <- bgcs %>%
   mutate(class = factor(class, levels = bigscape_classes), 
         class = my_classes[as.numeric(class)], 
         class = factor(class, levels = unique(my_classes))) %>%
  group_by(genome_id, class) %>%
  tally()
  
plot_tree <- ggtree(phylum_tree)

relative_widths <- phylum_tree_ds %>% 
  group_by(class) %>% 
  summarise(n = max(n)) %>% 
  pull(n)

bgcs_tree <- plot_tree

for (i in 1:length(unique(my_classes))){
  
  nbreaks <- relative_widths[i]/2
  title  <- my_classes[i]
  
  bgcs_tree <- bgcs_tree +
    geom_fruit(data = phylum_tree_ds[phylum_tree_ds$class == my_classes[i],], 
               geom = "geom_bar", 
               mapping = aes(y = genome_id, 
                             x = n, 
                             group = label), 
               pwidth = 0.015 * relative_widths[i],
               stat = "identity", 
               orientation = "y",
               fill = my_colors[i], 
               axis.params = list(axis = "x",
                                  text.size = 3,
                                  vjust = 1,
                                  nbreak = nbreaks,
                                  line.size = NA),
               grid.params = list(vline = FALSE)) 
}

families_df <- as_tibble(phylum_tree) %>% 
  filter(str_detect(label, 'f__')) %>% 
  mutate(label = str_replace(label, "-", ";"))

families_df$leaves <- sapply(
  lapply(
  families_df$label, 
  grepl, 
  x = genomes$gtdb_taxonomy), 
  sum)

families_df <- filter(families_df, leaves > 5)

annotated_tree <- bgcs_tree  +
  geom_highlight(data = families_df,
                 mapping = aes(node = node), 
                 fill = NA,
                 color = "lightgray",
                 extend = 10) +
  theme(legend.position = "none", 
        strip.background = element_rect(color = "gray95"), 
        strip.text = element_text(size = 10)) 

collapsed_tree <- annotated_tree

for (i in 1:nrow(families_df)) {
  collapsed_tree <- collapse(
    collapsed_tree, 
    clade_name = families_df$label[i], 
    node = families_df$node[i], 
    mode = "mixed", 
    color = "black", 
    fill = "gray95") }

clades_tree <- collapsed_tree  +
  geom_cladelab(data = families_df ,
                mapping = aes(node = node, label = label),
                barsize = NA, 
                geom = 'text', 
                colour = 'grey',
                bg.colour = 'white', 
                align = TRUE,
                offset = 0.3,
                hjust = 1,
                bg.r = 0.1,
                size =1.5)

clades_tree

```

Absence / Presence matrix for f\_\_Nitrospiraceae

```{r echo=FALSE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}

family <- "f__Nitrospiraceae"

family_ds <- 
  filter(genomes, str_detect(gtdb_taxonomy, family)) %>%
  mutate(AS_hqMAG = if_else(ncbi_bioproject == "PRJNA629478", TRUE, FALSE)) %>%
  filter(source == "refseq" |  AS_hqMAG) %>%
  mutate(species = str_extract(gtdb_taxonomy, "s__.*"), 
         species = str_sub(species, start = 4),
         tip_label = paste(genome_id, species))

family_tree <- tree_subset(
  phylum_tree,
  node = families_df$node[str_detect(families_df$label, family)], 
  levels_back = 0)

tips_to_drop <- family_tree$tip.label[
  family_tree$tip.label %in% family_ds$genome_id == FALSE]

family_tree <- drop.tip(family_tree, tips_to_drop)

data_tree <- left_join(family_tree,family_ds,
                         by = c("label" = "genome_id"))

matrix_ds <- bgcs %>%
  filter(genome_id %in% family_tree$tip.label) %>%
  mutate(GCF = str_split(GCF, ";")) %>%
  unnest(GCF) %>%
    mutate(class = my_classes[match(class, bigscape_classes)], 
           class = factor(class, levels = unique(my_classes))) %>%
  mutate(string = paste0(GCF, ": ", product))  %>%
  group_by(string) %>%
  mutate(singleton = n() == 1)

matrix_ds <- expand_grid(
  unique(matrix_ds$string), 
  unique(matrix_ds$genome_id)) %>%
  setNames(c("string", "genome_id")) %>%
  left_join(matrix_ds, by = c("string", "genome_id")) %>%
  mutate(class = factor(class, levels = unique(my_classes))) %>%
  arrange(class, product) %>%
  mutate(string = factor(string, levels = unique(string))) 
 
tree_matrix <-  ggtree(data_tree) +

    geom_fruit(
  geom = "geom_point", 
  data = matrix_ds, 
  mapping = aes(
    fill = class,
    x = string, 
    y = genome_id, 
    stroke = singleton), 
  size = 1.5, 
  offset = 0.5,
  shape = 21, 
  pwidth = 6,
  axis.params = list(axis = "x", 
                     text.size = 2.5, 
                     text.angle = 90, 
                     hjust =1),
  grid.params = list(vline = TRUE, 
                     size = 0.5, 
                     color = "gray95")) +
  
  geom_tiplab(mapping = aes(
    label = tip_label, 
    color = AS_hqMAG), 
              size = 1.7, 
              align = TRUE) + 
    scale_color_manual(values = c("black", "darkred")) +
  
  scale_fill_manual(values = my_colors, 
                    na.value = "gray95")  +
  scale_y_continuous(expand = c(0, 10, 0, 0)) +
  scale_shape_manual(values = c(NA, 15)) +
 guides(shape = FALSE,
        fill = guide_legend(nrow = 1, 
                            label.theme = element_text(size = 7),
                            title = ""), 
        color = FALSE)  +
  theme(legend.position = "top", 
        legend.direction = "horizontal",
        legend.box.spacing = unit(x=c(0,0,0,0),units="mm"))

tree_matrix

ggsave("figures/nitro_tree.svg", 
       tree_matrix,
       width = 9, 
       device = "svg")

source("notebook/remove_text_length.R")

remove_text_length("figures/nitro_tree.svg")
```

```{r}

barplot_ds <- bgcs %>% 
  filter(genome_id %in% family_ds$genome_id) %>%
  mutate(class = case_when(
    str_detect(product, "hserlactone") ~  "h-Ser-lactone", 
    str_detect(product, "phosphonate") ~ "Phosphonate", 
    TRUE ~ class)) %>%
  group_by(genome_id, class) %>% tally() %>%
  arrange(n)

classes_barplot <- c("Terpene", "RiPPs", "NRPS", "PKS-NRP_Hybrids", "PKSI", "Phosphonate", "h-Ser-lactone", "Others")

barplot_ds <- mutate(
  barplot_ds, 
  class = factor(class, levels = classes_barplot)) %>%
  arrange(class)

relative_widths <- barplot_ds %>% 
  group_by(class) %>%
  summarise(n = max(n)) %>% 
  pull(n)

barplot_colors <- c(my_colors[1:5], "cyan", "yellow", "grey")
bgcs_tree <- ggtree(data_tree) +
  
  geom_tiplab(mapping = aes(
    label = species, 
    color = AS_hqMAG), 
              size = 1.7, 
              align = TRUE) + 
    scale_color_manual(values = c("black", "darkred"))

for (i in 1:length(classes_barplot)){
  
 # nbreaks <- relative_widths[i]/2
  title  <- classes_barplot[i]
  class <- classes_barplot[i]
  nbreaks <- relative_widths[i]
    
  bgcs_tree <- bgcs_tree +
    geom_fruit(data = barplot_ds[barplot_ds$class == class,], 
               geom = "geom_bar", 
               mapping = aes(y = genome_id, 
                             x = n, 
                             group = label), 
               pwidth = 0.1 * relative_widths[i],
               stat = "identity", 
               orientation = "y",
               fill = barplot_colors[i], 
               axis.params = list(axis = "x",
                                  text.size = 3,
                                  vjust = 1,
                                  nbreak = nbreaks,
                                  line.size = NA),
               grid.params = list(vline = FALSE)) 
}

filename <- bgcs_tree
ggsave(bgcs_tree, filename = filename, height = 7, width = 9)



```
